### ISSUE TITLE
Token-Masking

### ISSUE THREAD

**Issue by Khanjan25** [IBM/zero-shot-classification-boost-with-self-training#2]
Can you please include the code for **token_masking** also? It seems to be really a good approach. There is a big improvement for ISAER especially.

**Comment by arielge** [comment]
Hi @Khanjan25,
Indeed we have not had the chance to integrate the masking into the repository.
But below you can see an example of the word masking approach that was used in the paper

```
import re
import string
import threading
from collections import defaultdict
from typing import List

import spacy
import numpy as np
from sklearn.neighbors import NearestNeighbors


spacy_models = defaultdict(lambda: None)
spacy_model_lock = threading.Lock()


STOP_WORDS = ["a", "about", "above", "after", "again", "against", "ain", "all", "am", "an", "and", "any",
              "are", "aren", "aren't", "as", "at", "be", "because", "been", "before", "being", "below",
              "between", "both", "but", "by", "can", "couldn", "couldn't", "d", "did", "didn", "didn't",
              "do", "does", "doesn", "doesn't", "doing", "don", "don't", "down", "during", "each", "few",
              "for", "from", "further", "had", "hadn", "hadn't", "has", "hasn", "hasn't", "have", "haven",
              "haven't", "having", "he", "her", "here", "hers", "herself", "him", "himself", "his", "how",
              "i", "if", "in", "into", "is", "isn", "isn't", "it", "it's", "its", "itself", "just", "ll",
              "m", "ma", "me", "mightn", "mightn't", "more", "most", "mustn", "mustn't", "my", "myself",
              "needn", "needn't", "no", "nor", "not", "now", "o", "of", "off", "on", "once", "only", "or",
              "other", "our", "ours", "ourselves", "out", "over", "own", "re", "s", "same", "shan", "shan't",
              "she", "she's", "should", "should've", "shouldn", "shouldn't", "so", "some", "such", "t",
              "than", "that", "that'll", "the", "their", "theirs", "them", "themselves", "then", "there",
              "these", "they", "this", "those", "through", "to", "too", "under", "until", "up", "ve", "very",
              "was", "wasn", "wasn't", "we", "were", "weren", "weren't", "what", "when", "where", "which",
              "while", "who", "whom", "why", "will", "with", "won", "won't", "wouldn", "wouldn't", "y",
              "you", "you'd", "you'll", "you're", "you've", "your", "yours", "yourself", "yourselves",
              "could", "he'd", "he'll", "he's", "here's", "how's", "i'd", "i'll", "i'm", "i've", "let's",
              "ought", "she'd", "she'll", "that's", "there's", "they'd", "they'll", "they're", "they've",
              "we'd", "we'll", "we're", "we've", "what's", "when's", "where's", "who's", "why's", "would"]


def remove_stop_words_and_punctuation(sentences: List[str]) -> List[str]:
    # remove punctuation
    sentences = remove_punctuation(sentences)
    # remove stop words
    regex = r"\b(" + "|".join(STOP_WORDS) + r")\b"
    sentences = [re.sub(regex, r"", text) for text in sentences]
    # remove extra spaces
    sentences = [' '.join(sent.split()) for sent in sentences]
    return sentences


def remove_punctuation(sentences: List[str]) -> List[str]:
    punctuation = string.punctuation + '•●'
    sentences = [t.translate(t.maketrans(punctuation, ' ' * len(punctuation))) for t in sentences]
    return sentences


class GloveMasker:
    def __init__(self, unk_token, n_to_mask=1, min_n_of_unmasked_words=3, algorithm='ball_tree'):
        self.unk_token = unk_token
        self.min_n_of_unmasked_words = min_n_of_unmasked_words
        self.algorithm = algorithm
        self.n_neighbors = n_to_mask
        self.spacy_model = self.get_spacy_model('en_core_web_lg')

    def get_spacy_model(self, spacy_model_name):
        # The model used for calculating the representations is loaded once, on the first time this method is called. On
        # subsequent calls, the loaded model is read from the spacy_models dictionary
        with spacy_model_lock:
            if spacy_models[spacy_model_name] is None:
                spacy_models[spacy_model_name] = spacy.load(spacy_model_name)  # type: ignore

        return spacy_models[spacy_model_name]

    def get_glove_representation(self, sentences: List[str]) -> List[np.ndarray]:
        """
        Given a list of texts, return a list of GloVe-based representation vectors. Each text is represented by the mean of
        the vector representations of its consisting tokens.
        :param sentences:
        :return: a list of numpy vectors. The vector length depends on the representation model specified under *language*
        """
        sentences = remove_stop_words_and_punctuation(sentences)
        # remove out-of-vocabulary tokens
        sentences = [' '.join(token for token in sent.split() if self.spacy_model.vocab.has_vector(token))
                     for sent in sentences]
        # the vector obtained by *make_doc(X).vector* is an average of the representations for the individual tokens in X
        embeddings = [self.spacy_model.make_doc(sent).vector for sent in sentences]  # type: ignore
        # logging.info(f"Done getting GloVe representations for {len(embeddings)} sentences")
        return embeddings

    def get_idxs_and_dists_of_closest_words_to_classnames(self, texts, classes):
        indices_of_closest_words = []
        dists_to_closest_words = []
        for text, classname in zip(texts, classes):
            word_vectors = [self.spacy_model.vocab.get_vector(t) if t not in STOP_WORDS else [-100]*300
                            for t in text.split()]
            
            cur_n_neighbors = min(self.n_neighbors,
                                  len(text.split())-self.min_n_of_unmasked_words)  # leaving at least two words unmasked

            if cur_n_neighbors > 0:
                nns = NearestNeighbors(n_neighbors=cur_n_neighbors, algorithm=self.algorithm).fit(word_vectors)
                class_emb = self.get_glove_representation([classname])
                indices_of_closest_words.append(nns.kneighbors(class_emb)[1][0])
                dists_to_closest_words.append( nns.kneighbors(class_emb)[0][0])
            else:
                indices_of_closest_words.append([])
                dists_to_closest_words.append([])

        return indices_of_closest_words, dists_to_closest_words
    
    def mask_texts(self, texts, classes):
        # get the indexes of the closest words for each text and classname
        indices_of_closest_words, _ = self.get_idxs_and_dists_of_closest_words_to_classnames(texts, classes)
            
        # mask the words appearing in the idxs from the above 
        texts_with_unk = []
        for text, similar_idxs in zip(texts, indices_of_closest_words):
            text_with_unk = []
            for idx, token in enumerate(text.split()):
                text_with_unk.append(self.unk_token if idx in similar_idxs else token)
            texts_with_unk.append(' '.join(text_with_unk))

        return texts_with_unk


def get_unk_token(model_name):
    if 'roberta' in model_name or 'bart' in model_name:
        unk = '<unk>'
    elif 'deberta' in model_name:
        unk = '[UNK]'
    else:
        raise NotImplementedError('Only bart, roberta and deberta models are currently supported')
    return unk


if __name__ == '__main__':
    model_name = 'roberta-large-mnli'
    masker = GloveMasker(unk_token=get_unk_token(model_name))

    texts = ['I had a great time at the park.', 'The movie made me very emotional.']
    target_class_names = ['joy', 'sadness']

    masked_texts = masker.mask_texts(texts=texts, classes=target_class_names)
    print(target_class_names)
    print(masked_texts)
```
